lib/common/socket.c

```diff
@@ -1570,59 +1570,196 @@ void h2o_sliding_counter_stop(h2o_sliding_counter_t *counter, uint64_t now)
+}
+
+static int ebpf_map_delete(int fd, const void *key)
+{
+    union bpf_attr attr = {
+        .map_fd = fd,
+        .key = (uint64_t)key,
+    };
+    return syscall(__NR_bpf, BPF_MAP_DELETE_ELEM, &attr, sizeof(attr));
+}
+
+static int file_exist(const char *path)
+{
+    struct stat s;
+    return stat(path, &s) == 0;
+}
+
+static int return_map_fd = -1; // for h2o_return
+
+int h2o_socket_ebpf_setup(void)
+{
+    const struct {
+        int type;
+        uint32_t key_size;
+        uint32_t value_size;
+    } map_attr = {
+        .type = BPF_MAP_TYPE_LRU_HASH,
+        .key_size = sizeof(pid_t),
+        .value_size = sizeof(uint64_t),
+    };
+
+    int success = 0;
+    int fd = -1;
+    if (getuid() != 0) {
+        h2o_error_printf("failed to set up eBPF maps because bpf(2) requires root privileges\n");
+        goto Exit;
+    }
+
+    if (!file_exist(H2O_EBPF_RETURN_MAP_PATH)) {
+        // creates a map and pinns the map to BPF fs
+        fd = ebpf_map_create(map_attr.type, map_attr.key_size, map_attr.value_size, H2O_EBPF_RETURN_MAP_SIZE,
+                             H2O_EBPF_RETURN_MAP_NAME);
+        if (fd < 0) {
+            if (errno == EPERM) {
+                h2o_error_printf("BPF_MAP_CREATE failed with EPERM, "
+                                 "maybe because RLIMIT_MEMLOCK is too small.\n");
+            } else {
+                h2o_perror("BPF_MAP_CREATE failed");
+            }
+            goto Exit;
+        }
+        if (ebpf_obj_pin(fd, H2O_EBPF_RETURN_MAP_PATH) != 0) {
+            if (errno == ENOENT) {
+                h2o_error_printf("BPF_OBJ_PIN failed with ENOENT, "
+                                 "because /sys/fs/bpf is not mounted as a BPF filesystem.\n");
+            } else {
+                h2o_perror("BPF_OBJ_PIN failed");
+            }
+            goto Exit;
+        }
+        success = 1;
+    } else {
+        // the pinned map file already exist
+        fd = ebpf_obj_get(H2O_EBPF_RETURN_MAP_PATH);
+        if (fd < 0) {
+            h2o_perror("BPF_MAP_CREATE failed");
+            goto Exit;
+        }
+        // make sure the critical attributes (type, key size, value size) are correct.
+        // otherwise usdt-selective-tracing does not work.
+        struct bpf_map_info m;
+        if (ebpf_obj_get_info_by_fd(fd, &m) != 0) {
+            h2o_perror("BPF_OBJ_GET_INFO_BY_FD failed");
+            goto Exit;
+        }
+        if (m.type != map_attr.type) {
+            h2o_error_printf("Unexpected map type: expected %d but got %d\n", map_attr.type, m.type);
+            goto Exit;
+        }
+        if (m.key_size != map_attr.key_size) {
+            h2o_error_printf("Unexpected map key size: expected %" PRIu32 " but got %" PRIu32 "\n", map_attr.key_size, m.key_size);
+            goto Exit;
+        }
+        if (m.key_size != map_attr.key_size) {
+            h2o_error_printf("Unexpected map value size: expected %" PRIu32 " but got %" PRIu32 "\n", map_attr.value_size,
```

Comment:
Regarding these error messages, I think we might be clear that the "map" we are talking about is the specific ebpf map (i.e. H2O_EBPF_RETURN_MAP_PATH) that we are dealing with. Otherwise, "map" is such a generic term that users might have difficulty in finding out the cause.